### PR TITLE
enhancement(splunk_hec sink, humio_logs sink, codecs): Integrate `encoding::Encoder` with `splunk_hec`/`humio_logs` sink

### PR DESCRIPTION
**Pull request by pablosichert**
Part of #9459.

### CODE REVIEW

**Review by jszwedko — APPROVED** [review]
Glad to see this cleaned up!

File: src/sinks/splunk_hec/logs/encoder.rs [hunk 1]
```diff
@@ -95,30 +60,40 @@ impl Encoder<Vec<HecProcessedEvent>> for HecLogsEncoder {
...
+                let metadata = processed_event.metadata;
+                self.transformer.transform(&mut event);
+
+                let hec_event = if matches!(encoder.serializer(), Serializer::Json(_)) {
+                    HecEvent::Json(event.into_log())
+                } else {
+                    let mut bytes = BytesMut::new();
+                    encoder.encode(event, &mut bytes).ok()?;
+                    HecEvent::Bytes(bytes.freeze())
+                };
```
Comment (pablosichert) [inline src/sinks/splunk_hec/logs/encoder.rs:77]: I'm wondering if we want to ship `Serializer::NativeJson` as JSON or JSON-encoded string here.  Do you have thoughts on that, @lukesteensen @jszwedko?

Comment (jszwedko) [reply src/sinks/splunk_hec/logs/encoder.rs:77]: We should just match whatever the `json` encoding was previously doing.

Comment (pablosichert) [reply src/sinks/splunk_hec/logs/encoder.rs:77]: > We should just match whatever the json encoding was previously doing.

We do that in the `Serializer::Json` case here. I was interested specifically in the `NativeJson` case which is supposed to represent our internal events, so may differ from the `Json` codec in the future

Comment (jszwedko) [reply src/sinks/splunk_hec/logs/encoder.rs:77]: Ah, yes, sorry, I should have been clearer. I think we should encode it the same way as we encode `Json` to send it to Splunk. In this case, it should encode into `HecEvent::Json()`.

Comment (pablosichert) [reply src/sinks/splunk_hec/logs/encoder.rs]: I moved that logic into `Serializer::supports_json`/`Serializer::encode_json`:

https://github.com/vectordotdev/vector/blob/fa304f628d71145182a57d58bff228ef086118d5/src/sinks/splunk_hec/logs/encoder.rs#L72-L84

File: lib/codecs/src/encoding/format/json.rs [hunk 2]
```diff
@@ -35,6 +35,11 @@ impl JsonSerializer {
     pub const fn new() -> Self {
         Self
     }
+
+    /// Encode event as JSON.
+    pub fn encode_json(&self, event: Event) -> Result<serde_json::Value, serde_json::Error> {
+        event.try_into()
```
Comment (lukesteensen) [inline lib/codecs/src/encoding/format/json.rs:41]: This doesn't seem like it's going to match the structure of the `encode` method below, but will include the `Event` wrapper that's used in `NativeJsonSerializer`.

Having these kinds of similar-but-different methods bring the danger of these types of bug, so it'd be nice to at least see some tests that the outputs are equivalent.

File: lib/codecs/src/encoding/mod.rs [hunk 3]
```diff
@@ -241,6 +241,32 @@ pub enum Serializer {
...
+        }
+    }
+
+    /// Encode event as JSON.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the serializer does not support encoding to JSON. Call `Serializer::supports_json`
+    /// if you need to determine the capability to encode to JSON at runtime.
+    pub fn encode_json(&self, event: Event) -> Result<serde_json::Value, serde_json::Error> {
```
Comment (lukesteensen) [inline lib/codecs/src/encoding/mod.rs:259]: This name is somewhat confusing. I might call it something like `to_json_value` or similar to differentiate it from normal encoding.

File: lib/codecs/src/encoding/format/json.rs (hunk 2 above)
Comment (pablosichert) [reply lib/codecs/src/encoding/format/json.rs]: Oh damn, that's a great catch!

Fixed and added tests in 0cba4ba9c78fed9e0d0735eed008771dc2ee8a42

Comment (pablosichert) [reply lib/codecs/src/encoding/format/json.rs]: Oh damn, that's a great catch!

Fixed and added tests in 0cba4ba9c78fed9e0d0735eed008771dc2ee8a42

File: lib/codecs/src/encoding/mod.rs (hunk 3 above)
Comment (pablosichert) [reply lib/codecs/src/encoding/mod.rs]: Done with c0bcf7d3ca5e1c54c575a39f6494ffd05feac0dc

**Review by lukesteensen — APPROVED** [review]

**Pull request merged**